<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2026-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Joel Gómez Uribe
### Rodrigo Vega Vilchis
---

## 1.1.3. Estructuras de control, funciones y bibliotecas

En este *Notebook* presentaremos de manera muy breve el uso general de las estrcuturas de control en *Python*, así también, cómo definir y usar funciones dentro de *Python*. Asimismo, mostraremos muy brevemente, cómo leer y escribir archivos (de texto o de datos) y finalmente expondremos cómo crear _**scripts** de Python_.

### Estructuras de control

---

En *Python* existen las estructuras de control comunes usadas en los lenguajes programación, que son las condicionales (`if`) y las iteretivas (`for` y `while`).

Estás se usan junto con, los operadores lógicos y de comparación ya mencionados y que tienen un formato en común en casi todos los lenguajes de programación.

#### Operadores de comparación:

- *Igual que*:  `==`
- *Diferente que*:  `!=`
- *Menor que*:  `<`
- *Mayor que*:  `>`
- *Menor o igual que*:  `<=`
- *Mayor o igual que*:  `>=`

#### Operadores de lógicos:

- *AND*:  `and`
- *OR*:  `or`
- *NOT*:  `!`

A diferencia de otros lenguajes de programación en *Python*, no existen instrucciones que indiquen donde empieza y termina un bloque del programa.

Un bloque empieza por **`:`**, y su extensión se indica por el uso de una cantidad definida de espacio en blanco al principio de cada línea. 

Es muy importante dejar **una identación (un tabulador)** después de que se inicia un bloque de instrucciones definido por `:`.

Algunos editores de texto (como Kate) o IDEs permiten hacer eso de forma automática.

### Estructuras Condicionales

Para checar una condición y ejecutar código dependiente del resultado, se ocupa  **`if`**:

In [1]:
a = int(input('Ingresa el valor de a '))

if (a > 0):
        print("a es positiva")
        a = a + 1
elif a == 0:  # Notese: dos veces =
        print("a es 0")
else:
        print("a es negativa")

Ingresa el valor de a  1


a es positiva


En el primer caso, hay dos comandos en el bloque que se ejecutará en el caso de que `a > 0`. Nótese que `elif` es una abreviación de `else:if`

Las condiciones que se pueden ocupar incluyen: `<`, `<=` y `!=` (*no es igual*). 
Para combinar condiciones se ocupan las palabras `and` y `or`:

In [2]:
a = 3; b=7

if a > 1 and b > 2:
    print("Grandes")

if a > 0 or b > 0:
    print("Al menos uno positivo")

Grandes
Al menos uno positivo


Aca otro ejemplo

In [3]:
x = int(input("Introduce un número entero no mayor que diez: "))

if x > 10:
    print("Ingresaste un número mayor a diez.")
    print("Déjame arreglar eso por ti.")
    x = 10
    print("tu número es",x)

Introduce un número entero no mayor que diez:  9


Donde podemos usar `and` y `or`:

In [4]:
x = int(input("Introduce un número entero entre 1 y diez: "))
    
if x > 10 or x < 1:
    print("Tu número es demasiado grande o demasiado pequeño.")


Introduce un número entero entre 1 y diez:  5


In [5]:
x = int(input("Introduce un número entero menor o igual a 1 y menor o igual a diez: "))
    
if x <= 10 and x >=1:
    print("Tu número es correcto.")
else:
    print("Tu número es incorrecto.")

Introduce un número entero menor o igual a 1 y menor o igual a diez:  1


Tu número es correcto.


In [6]:
x = int(input("Introduce un número entero no mayor que diez: "))
if x > 10:
    print("Tu numero es mayor que diez.")
elif x > 9:
    print("Tu número está bien, pero por poco no!")
else:
    print("Tu número está bien. Continua.")

Introduce un número entero no mayor que diez:  8


Tu número está bien. Continua.


**Ejemplo: Número pares e impares**

Escribamos un programa que tome como entrada un número entero e imprima si el número es par o impar.

In [7]:
n = int(input("Ingresa un entero: "))
if n%2==0:
    print("par")
else:
    print("impar")

Ingresa un entero:  5000


par


## Estructuras Iterativas

La parte central de muchos cálculos científicos consiste en llevar a cabo **ciclos (bucles)**. Para esto usamos las estructura de control **`while`** y **`for`**.

La instrucción `while` ejecuta un bloque de código mientras una cierta condición se satisfaga. Asimismo, la instrucción `for` se usa generalmente cuando se conoce de antemano el número de iteraciones necesarias.

### La estructura `for`

Un ciclo tipo `for` se suele ocupar para llevar a cabo un número de iteraciones que se conoce de antemano. En el bucle de tipo `for`, encontramos una diferencia importante con respecto a lenguajes más tradicionales:

Se puede iterar sobre cualquier lista y ejecutar un bloque de código para cada elemento de la lista:

In [8]:
r = [ 1, 3, 5]

for n in r:
    print(n)
    print(2*n)
print("Termine!")

1
2
3
6
5
10
Termine!


In [9]:
l = [1, 2.5, -3.71, "hola", [2, 3]]

for i in l:
    print(2*l)

[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]
[1, 2.5, -3.71, 'hola', [2, 3], 1, 2.5, -3.71, 'hola', [2, 3]]


#### La función *range()*

Para averiguar que es lo que hace una función, se puede investigar de manera interactiva:

In [10]:
range?

Init signature: range(self, /, *args, **kwargs)
Docstring:     
range(stop) -> range object
range(start, stop[, step]) -> range object

Return an object that produces a sequence of integers from start (inclusive)
to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
These are exactly the valid indices for a list of 4 elements.
When step is given, it specifies the increment (or decrement).
Type:           type
Subclasses:     

`range(N)` en `python 2.x` devuelve una lista que contiene una progresión aritmética de enteros. Sin embargo en `python 3.x`,`range` es una función que nos devuelve un rango de números naturales a los cuales podemos acceder por medio de un iterador.

In [11]:
range(10)

range(0, 10)

Si queremos iterar sobre muchos elementos, es más útil construir la lista con `range`.

In [12]:
r = range(5)
print("Hola!")

for n in r:
    print("Hola de nuevo!")

Hola!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!
Hola de nuevo!


 Por ejemplo, para hacer una iteración para todos los números pares hasta 100, podemos utilizar:

In [13]:
for i in range(100):
    print(2*i)

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48
50
52
54
56
58
60
62
64
66
68
70
72
74
76
78
80
82
84
86
88
90
92
94
96
98
100
102
104
106
108
110
112
114
116
118
120
122
124
126
128
130
132
134
136
138
140
142
144
146
148
150
152
154
156
158
160
162
164
166
168
170
172
174
176
178
180
182
184
186
188
190
192
194
196
198


Recordemos que la función `range` crea una lista de `N` elementos comenzado en el *$0$*. O también podemos usar:

In [14]:
for i in range(3,57,3):
    print(2*i)

6
12
18
24
30
36
42
48
54
60
66
72
78
84
90
96
102
108


Para poder indicar **`un inicio`**, **`un fin`** y **`un incremento`** de paso en la iteración.

`range()` también es util para llenar listas:

In [15]:
L = []
print(L)
for i in range(1,10,3):
    L.append(i)
    print(L)

[]
[1]
[1, 4]
[1, 4, 7]


Podemos resumir el código anterior en una sola instrucción.

Por ejemplo, para obtener una _lista_ con los diez primeros números podemos utilizar un `for` (lo explicaremos más adelante) de la siguiente manera:

In [16]:
L1 = [i for i in range(10)]
L1

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Nótese que la lista comienza en 0 (cero) y que **se omite el elemento final**., 

Por ejemplo para obtner una lista de cuatro elementos que llegue hasta el **3** se puede hacer lo siguiente:

In [17]:
[i for i in range(4)]

[0, 1, 2, 3]

También se le puede indicar en donde inicia y donde termina la lista:

In [18]:
[i for i in range(3,10)]

[3, 4, 5, 6, 7, 8, 9]

O bien también se le puede indicar un paso de incremento:

In [19]:
[i for i in range(3,17,2)]

[3, 5, 7, 9, 11, 13, 15]

In [20]:
[i for i in range(3,17,3)]

[3, 6, 9, 12, 15]

In [21]:
[i for i in range(3,17,5)]

[3, 8, 13]

Sobra decir que estas lista generadas por `range()` las podemos almacenar en una variable.

In [22]:
x = [i for i in range(3,17,5)]
print(x)

[3, 8, 13]


In [1]:
import numpy as np


### **Ejemplo:** Lineas de emisión de Hidrogeno 

En 1888 Johannes Rydberg publicó su famosa fórmula para las longitudes de onda $\lambda$ de las líneas de emisión del átomo de hidrógeno:

<h3 align="center">$\displaystyle \dfrac{1}{\lambda} = R \left(\dfrac{1}{m^2} - \dfrac{1}{n^2} \right)$</h3> 

donde $R$ es la constante de Rydberg $R = 1.097 × 10^{-2}$ nm$^{-1}$ y $m$ y $n$ son números enteros positivos. Para un valor dado de $m$, las longitudes de onda $\lambda$ dadas por esta fórmula para todo $n > m$ forman una serie, las tres primeras de tales series, para $m = 1$, $2$ y $3$, se conocen como series de *Lyman*, *Balmer* y *Paschen* en honor a sus respectivos descubridores.

Para resumir varias de las cosas vistas, a continuación escribiremos un breve programa de ejemplo, que imprime las longitudes de onda de las primeras cinco líneas en cada una de estas tres series:


In [23]:
R = 1.097e-2

for m in range(1,4):    
    print("\n Serie para m =",m)    
    for n in range(m+1,m+6):
        invlambda = R*(1/m**2-1/n**2)
        print("  ",1/invlambda,"nm")



 Serie para m = 1
   121.5436037678517 nm
   102.55241567912488 nm
   97.23488301428137 nm
   94.95594044363415 nm
   93.76220862091418 nm

 Serie para m = 2
   656.3354603463993 nm
   486.1744150714068 nm
   434.084299170899 nm
   410.2096627164995 nm
   397.04243897498225 nm

 Serie para m = 3
   1875.2441724182836 nm
   1281.9051959890612 nm
   1093.8924339106654 nm
   1005.013673655424 nm
   954.6697605038536 nm


También nótese que `range()` **no acepta** argumentos de punto **flotante**.

In [24]:
[i for i in range(3.,10.)]

TypeError: 'float' object cannot be interpreted as an integer

Recordemos también que `range` sólo funciona con enteros. Por lo que el siguiente programa nos arrojará un error:

In [ ]:
p = 10
q = 3
for n in range(p/q):
    print(n)

Para poder corregir lo anterior, podemos usar la parte entera:

In [ ]:
for n in range(p//q):
    print(n)

O bien la función `arange` de **`numpy`**.

In [ ]:
import numpy as np

for n in np.arange(p/q):
    print(n)

### La estructura `while`


Como instrucción de bucle, la más simple es **`while`**, que como ya mencionamos, ejecuta un bloque de código mientras una cierta condición se satisfaga, y suele ocuparse para llevar a cabo una iteración en la cual no se conoce de antemano el número de iteraciones necesarias. Sin embargo hay que tener cuidado al usar esta estructura pues si no definimos bien la condición de termino puede que el ciclo se repita indefinidamente y nunca acabe. 

Un ejemplo muy sencillo de un while, donde sí se conoce la condición terminal, para contar hasta 9, se puede hacer como sigue:

In [ ]:
i = 0
while i < 10:
    # print(i)
    i += 1     # equivalente a i = i + 1
    print(i)

O bien, podemos usarlo de la siguiente manera:

In [ ]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x > 10:
    
    print("Este numero es mayor a diez. Vuelve a intentarlo.")
    x = int(input("Ingresa un número entero no mayor a diez: "))

print("Tu número es",x)

Ahora compiquemos un poco el programa y de los números par e impar y pidamos uno y uno:

In [ ]:
print("Ingresa dos enteros, uno par y el otro impar.")

m = int(input("Ingresa el primer entero: "))
n = int(input("Ingresa el segundo entero: "))

while (m+n)%2==0:
    print("Uno debe ser par y el otro impar.")
    
    m = int(input("Ingresa el primer entero: "))
    n = int(input("Ingresa el segundo entero: "))
    
print("Correto! Los números que elegiste son:",m,"y",n)

### **Ejemplo: Sucesión de Fibonacci**

Los números de (la sucesión) de Fibonacci se definen de la siguiente manera:

### $f_n = f_{n-1} + f_{n-2}$

con las condiciones iniciales

#### $f_0=0$  y $f_1=1$

<!-- ![fibrabbits.png](attachment:fibrabbits.png) -->
<div>
    <img src="fibrabbits.png", width="800" height="600">
</div>

Escribamos un programa para generar los números de Fibonacci, menores a 1000:

In [1]:
fn_1,fn_2 = 1,1
while fn_2 < 1000:    
    print(fn_2)
    fn_1,fn_2 = fn_2,fn_1+fn_2

1
2
3
5
8
13
21
34
55
89
144
233
377
610
987


In [4]:
fn_1,fn_2 = 1,1
while fn_2 < 10e10:    
    print(fn_2)
    fn_1,fn_2 = fn_2,fn_1+fn_2

1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765
10946
17711
28657
46368
75025
121393
196418
317811
514229
832040
1346269
2178309
3524578
5702887
9227465
14930352
24157817
39088169
63245986
102334155
165580141
267914296
433494437
701408733
1134903170
1836311903
2971215073
4807526976
7778742049
12586269025
20365011074
32951280099
53316291173
86267571272


In [ ]:
fn_1,fn_2 = 1,1
while fn_2 < 10:    
    print("fn-1=",fn_1,"fn-2=",fn_2)
    print("<---------")
    fn_1,fn_2 = fn_2,fn_1+fn_2
    print("fn-1=",fn_1,"fn-2=",fn_2)
    print("**********")

Una instrucción muy útil, cuando estamos usando un `while` es **`break`**, que nos permiten salir del ciclo incluso si no se cumple la condición. Por ejemplo:

In [2]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x>10:

    print("Esto es mayor a diez, por favor intentalo de nuevo.")
    x = x+1
    
    if x==111:
        break

    print(x)

Ingresa un número entero no mayor a diez:  1


Una instrucción parecida a `break`  es la instrucción **`continue`**. Al colocar esta instrucción dentro `while` (con una condición de un `if`), hará que el programa salte el resto del código del ciclo, regresando al principio del bucle, donde se vuelve a comprobar la condición en el `while`.

**OJO:** Se recomienda un uso bien pensando de esta istrucción, pues podría provocar que nuestro programa se estanque en un *"bucle infinito"*





In [ ]:
x = int(input("Ingresa un número entero no mayor a diez: "))

while x>10:
    
    print("Esto es mayor a diez, por favor intentalo de nuevo.")
    x = x+1
    
    if (x % 2) == 0:
        continue
    elif x==111:
            break
    print(x)


<!-- 
break | continue
- | - 
![break](break.png) | ![continue](continue.png)
 -->

### Diferencia entre el uso de `break`  y `continue`
$\;$

<div align="center">
    <table>      
        <tr>
            <td> <img src="break.png" alt="break" style="width: 450px;"/> </td>
            <td> <img src="continue.png" alt="continue" style="width: 450px;"/> </td>
        </tr>
        <tr>
            <td><div align="center"><h2>break</h2></div></td>
            <td><div align="center"><h2>continue</h2></div></td>
        </tr>          
    </table>
</div>



### **Ejercicio:** Números de Catalan

Los números de Catalan $C_n$ son una secuencia de números enteros 1, 1, 2, 5, 14, 42, 132 que juegan un papel importante en la mecánica cuántica y la teoría de los sistemas desordenados. (Eran fundamentales para la prueba de Eugene Wigner de la llamada ley del semicírculo). Se definen de la siguiente manera:

<h3 align="center">$C_0 = 1,\qquad C_{n+1} = {4n+2\over n+2}\,C_n$.</h3> 

Escribe un programa que imprima en orden creciente todos los números catalanes menores o iguales a mil millones.


In [6]:
C0 = 1
n = 0
x = C0

#Formula
C0 = (4*n+2)/(n+2)*C0

while C0 <= 10e8:
    print(C0)
    
    if x < 1:
        continue
    elif x == 1:
        break
    print(C0)

1.0


In [1]:
C0 = 1
for n in range(100):
    Cn = (4*n+2)/(n+2)*C0
    C0 = Cn
    print(Cn)

1.0
2.0
5.0
14.0
42.0
132.0
429.0
1430.0
4862.0
16796.0
58786.0
208012.0
742900.0
2674440.0
9694845.0
35357670.0
129644790.0
477638700.0
1767263190.0
6564120420.0
24466267020.0
91482563640.0
343059613650.0
1289904147324.0
4861946401452.0
18367353072152.0
69533550916004.0
263747951750360.0
1002242216651368.0
3814986502092304.0
1.4544636039226908e+16
5.55340648770482e+16
2.1233613041224314e+17
8.12944042149731e+17
3.116285494907302e+18
1.1959798385860456e+19
4.5950804324621754e+19
1.7673386278700676e+20
6.804253717299761e+20
2.6221270422764934e+21
1.0113918591637903e+22
3.904442991190447e+22
1.5085347920508544e+23
5.83300119592997e+23
2.2571178540772493e+24
8.740328711533179e+24
3.386877375719107e+25
1.3132789824216945e+26
5.0955224517961745e+26
1.9782616577561618e+27
7.684785670514321e+27
2.9869166945772643e+28
1.161578714557825e+29
4.519597180279538e+29
1.75941461660882e+30
6.852456927844878e+30
2.670095285677487e+31
1.0408846028912238e+32
4.059449951275773e+32
1.583850964596121e+33
6.

In [ ]:
C0 = 1
for n in range(100):
    Cn = (4*n+2)/(n+2)*C0
    C0 = Cn
    print(Cn)
    while Cn <= 10e8:
        print(Cn)
        
        if C0 < 1:
            continue
        elif C0 == 1:
            break
            print(Cn)

### **Ejercicio:** 

Escribe un programa que realice la siguiente suma:

<h4 align="center">$\displaystyle s = \sum_{k=1}^{100} \dfrac{1}{k}$</h4> 

In [5]:
N = 100
S = 0.0

for k in range(1, N+1):
    S = S + 1/k
S

5.187377517639621

### **Ejercicio:** 

Ahora escribe un programa que realice la suma de los cuadrados de los valores tomados de un archivo de texto:

<h4 align="center">$\displaystyle s = \sum x^2 \; \; \; \forall x \in V_{txt}$</h4> 

In [7]:
N = 100
S = 0.0

for k in range(1, N+1):
    S = S + x**2
    S += x**2
S

200.0

## Funciones

---
Las funciones se pueden considerar como subprogramas que ejecutan una tarea dada, pueden contener variables propias (locales) así como estructuras de control y cálculos en su interior. En *Python*, las funciones pueden o no aceptar argumentos, y pueden o no regresar resultados.

La sintaxis para declarar una función es como sigue:

In [25]:
def funcion(x):
    print("Argumento x = ", x)
    return x*x

y se llama así:

In [30]:
funcion(8)
funcion(3.5)

Argumento x =  8
Argumento x =  3.5


12.25

Las funciones se pueden utilizar con argumentos de cualquier tipo (el tipo de los argumentos nunca se especifica). Si las operaciones llevadas a cabo no se permiten para el tipo que se provee, entonces *Python* regresa un error:

In [31]:
funcion("hola")

Argumento x =  hola


TypeError: can't multiply sequence by non-int of type 'str'

Las funciones pueden regresar varios resultados al juntarlos en un *tuple* (n-ada):

In [ ]:
def f(x, y):
    return 2*x, 3*y

In [ ]:
f(3,4)

Se puede proporcionar una forma sencilla de documentación de una función al proporcionar un *docstring* (cadena de documentación):

In [ ]:
def cuad(x):
    """Funcion para elevar un numero al cuadrado.
    Funciona siempre y cuando el tipo de x permita la multiplicacion.
    """
    return x*x

Ahora si interrogamos al objeto `cuad` (con ipython o Jupyter):

In [ ]:
cuad?

nos da esta información. Si la función está definida en un archivo, entonces `cuad??` muestra el código de la definición de la función.

In [ ]:
cuad??

## BIBLIOTECAS.

---


Se puede escribir un archivo de texto (un *script*) que solo contenga definiciones de funciones, a la cual puede considerarse una biblioteca propia, y también se puede importarse:

`from mi_biblioteca import *`

Así podemos llamar a esta biblioteca desde el *interprete de comandos python* (o el entrono **ipython** o **Jupyter**) y usar sus funciones dentro de este.

O bien, se puede importar detro de otro *script*, con lo que podemos llamar esas funciones o ver sus características coma ya mencionamos.

In [ ]:
# !touch miBiblio.py

In [ ]:
import miBiblio as mb 

In [ ]:
mb.

In [ ]:
mb.funcion(9)

### **Ejercicio:** 

Escribe una función que tome como argumento un número entero y que regrese el factorial del mismo.

<h4 align="center">$\displaystyle n! = \prod_{i=1}^n i$</h4> 

Y pruebala para diferentes valores.

### **Ejercicio:** 

Supongamos, que tenemos un punto $p$ especificado en coordenadas cilíndricas $r$, $\theta$, $z$. Escribe una función que tome como argumentos los valores de las cordenadas del punto ($r$,$\theta$,$z$) para que regrese la distancia $d$ del origena a $p$ y pruebala para diferentes valores. 

*Hint:* La forma más sencilla es usar el teorema de Pitagoras.

In [1]:
def Disancia (r,theta,z):
    p = mt.sqrt(r**2 + z**2)
    return p

### **Ejercicio:** 

Supongamos que tenemos un número entero $n$ y queremos saber sus **_factores primos_**. Los factores primos se pueden calcular con relativa facilidad dividiendo repetidamente por todos los números enteros desde 2 hasta $n$ y verificando si el resto es cero. 

Escribe una función que tome como argumento un número $n$ y devuelva una lista de sus factores primos.